pkg/dequeuer/async_handler.go

```diff
@@ -0,0 +1,206 @@
+		updateStatusErr := h.updateStatus(requestID, async.StatusFailed)
+		if updateStatusErr != nil {
+			return errors.Wrap(updateStatusErr, fmt.Sprintf("failed to update status to %s", async.StatusFailed))
+		}
+		return nil
+	}
+
+	if err = h.uploadResult(requestID, result); err != nil {
+		updateStatusErr := h.updateStatus(requestID, async.StatusFailed)
+		if updateStatusErr != nil {
+			h.log.Errorw("failed to update status after failure to upload result", "id", requestID, "error", updateStatusErr)
+		}
+		return errors.Wrap(err, "failed to upload result to storage")
+	}
+
+	if err = h.updateStatus(requestID, async.StatusCompleted); err != nil {
+		return errors.Wrap(err, fmt.Sprintf("failed to update status to %s", async.StatusCompleted))
+	}
+
+	h.log.Infow("workload processing complete", "id", requestID)
+
+	return nil
+}
+
+func (h *AsyncMessageHandler) updateStatus(requestID string, status async.Status) error {
+	key := async.StatusPath(h.storagePath, requestID, status)
+	return h.aws.UploadStringToS3("", h.config.Bucket, key)
+}
+
+func (h *AsyncMessageHandler) getPayload(requestID string) (*userPayload, error) {
+	key := async.PayloadPath(h.storagePath, requestID)
+	output, err := h.aws.S3().GetObject(
+		&s3.GetObjectInput{
+			Key:    aws.String(key),
+			Bucket: aws.String(h.config.Bucket),
+		},
+	)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	contentType := "application/octet-stream"
+	if output.ContentType != nil {
+		contentType = *output.ContentType
+	}
+
+	return &userPayload{
+		Body:        output.Body,
+		ContentType: contentType,
+	}, nil
+}
+
+func (h *AsyncMessageHandler) deletePayload(requestID string) {
+	key := async.PayloadPath(h.storagePath, requestID)
+	err := h.aws.DeleteS3File(h.config.Bucket, key)
+	if err != nil {
+		h.log.Errorw("failed to delete user payload", "error", err)
+		telemetry.Error(errors.Wrap(err, "failed to delete user payload"))
+	}
+}
+
+func (h *AsyncMessageHandler) submitRequest(payload *userPayload, requestID string) (interface{}, error) {
+	req, err := http.NewRequest(http.MethodPost, h.config.TargetURL, payload.Body)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	req.Header.Set("Content-Type", payload.ContentType)
+	req.Header.Set(CortexRequestIDHeader, requestID)
+	response, err := h.httpClient.Do(req)
+	if err != nil {
+		return nil, ErrorUserContainerNotReachable(err)
+	}
+
+	defer func() {
+		_ = response.Body.Close()
+	}()
+
+	if response.StatusCode != http.StatusOK {
+		return nil, ErrorUserContainerResponseStatusCode(response.StatusCode)
+	}
+
+	if !strings.HasPrefix(response.Header.Get("Content-Type"), "application/json") {
+		return nil, ErrorUserContainerResponseMissingJSONHeader()
+	}
+
+	var result interface{}
+	if err = json.NewDecoder(response.Body).Decode(&result); err != nil {
+		return nil, ErrorUserContainerResponseNotJSONDecodable()
+	}
+
+	return result, nil
```

Comment:
Since it has been verified that response.Body is json, we can just return the bytes. That way, we don't need to marshal the result in uploadResult.

Reply:
I'm not following. I have to attempt to decode the result to json to check if it is valid json, and then I just return it to later upload it. How would we skip a step here?

Reply:
When it is being uploaded, I believe there is another json.marshal step that marshal's the result. If this function returned byte array, then you we can use `UploadBytesToS3` instead of `UploadJSONToS3`.